### PR TITLE
fix: harden username blocklist

### LINKED ISSUE #39282: Harden Username Blocklist
Reference: https://twitter.com/zersiax/status/1284098104711622661

A generic commonly accepted excludelist (serverside), and a profanity filter check on ~~client~~ server side using https://github.com/web-mech/badwords should cover most scenarios. 

Edit: Client side validation is bad! Sorry it was late when I added this requirement.

### PR DESCRIPTION
**Pull request by raisedadead**
- [x] Checks if a username is from a common list of paths, routes, terms
- [x] Checks requested username against a profanity filter

### CODE REVIEW

File: api-server/server/utils/constants.js [hunk 1]
```diff
@@ -83,5 +84,537 @@ export const blacklistedUsernames = [
...
+  'www2',
+  'www3',
+  'www4',
+  'you',
+  'yourname',
+  'yourusername',
+  'zlib'
 ];
+
+export const blocklistedUsernames = blocklist.filter((value, index, self) => {
```
Comment (RandellDawson) [inline api-server/server/utils/constants.js:618]: The current filter combined with indexOf  may not be as efficient as the following:
```suggestion
export const blocklistedUsernames = [ ...new Set(blocklist) ];
```

File: api-server/server/utils/constants.js [hunk 2]
```diff
@@ -83,5 +84,537 @@ export const blacklistedUsernames = [
...
+  return self.indexOf(value) === index;
+});
```
Comment (RandellDawson) [inline api-server/server/utils/constants.js:620]: The current filter combined with indexOf  may not be as efficient as the following:
```suggestion
export const blocklistedUsernames = [ ...new Set(blocklist) ];
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
Looks good.  See my minor suggestion.

**Comment by RandellDawson** [comment]
@raisedadead It appears the linter prefers the following?
```
export const blocklistedUsernames = [...new Set(blocklist)];
```

**Review by RandellDawson — DISMISSED** [review]

**Comment by nabraas60000** [comment]
json.filter(function(val)

File: api-server/common/models/user.js [hunk 3]
```diff
@@ -347,10 +348,14 @@ export default function(User) {
...
-
-    // check to see if username is on blacklist
-    if (username && blacklistedUsernames.indexOf(username) !== -1) {
+    log('check if username is available');
+    // check to see if username is on blocklist
+    const usernameFilter = new badwordFilter();
+    if (
+      username &&
+      (blocklistedUsernames.indexOf(username) !== -1 ||
+        usernameFilter.isProfane(username) === true)
```
Comment (ojeytonwilliams) [inline api-server/common/models/user.js:357]:
```suggestion
      (blocklistedUsernames.includes(username) ||
        usernameFilter.isProfane(username))
```
Nitpicky, I know, but `isProfane` returns `true` or `false`.

Also, `isProfane` only catches things like `profanewordone-profanewordtwo` not `profanewordoneprofanewordtwo`.  Do we want to deal with those manually?

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
I checked it out locally and it works nicely.  Nitpick below:

File: api-server/common/models/user.js (hunk 3 above)
Comment (raisedadead) [reply api-server/common/models/user.js:357]: > Do we want to deal with those manually?

Not sure how though, without doing so statically.

**Review by ojeytonwilliams — APPROVED** [review]

**Pull request merged**